### PR TITLE
Contract staking 

### PR DESCRIPTION
**Pull request by EdHastingsCasperAssociation**
This PR introduces support to delegate using a purse instead of a public key. This allows, among other things, a contract with an associated purse to delegate.

### CODE REVIEW

File: execution_engine/src/runtime/auction_internal.rs
Comment (darthsiroftardis) [inline execution_engine/src/runtime/auction_internal.rs:121]: So, is there a reason we piggyback the write logic to prune by passing None, would it make sense to move the logic out to a `prune_unbond`, feels like this is ripe for programmer error.

Comment (darthsiroftardis) [reply execution_engine/src/runtime/auction_internal.rs:121]: I get that we were already piggybacking the logic in the previous, my question is more about just separating the prune v. write concern

File: execution_engine_testing/test_support/src/wasm_test_builder.rs
Comment (darthsiroftardis) [inline execution_engine_testing/test_support/src/wasm_test_builder.rs]: Nit: Comment doesn't match the type

File: execution_engine/src/runtime/auction_internal.rs
Comment (EdHastingsCasperAssociation) [reply execution_engine/src/runtime/auction_internal.rs:121]: Well...that would make it the callers problem to deal with. So...sth like 7 call sites would need to duplicate the same check.

File: execution_engine_testing/test_support/src/wasm_test_builder.rs
Comment (EdHastingsCasperAssociation) [reply execution_engine_testing/test_support/src/wasm_test_builder.rs]: Fix will be on next commit.

File: execution_engine/src/runtime/auction_internal.rs
Comment (darthsiroftardis) [reply execution_engine/src/runtime/auction_internal.rs:121]: Fair

File: execution_engine/src/runtime/auction_internal.rs
Comment (darthsiroftardis) [inline execution_engine/src/runtime/auction_internal.rs:362]: Seems like the purse is being used to do a read only, is there a reason we are creating it with all the perms here??

Comment (darthsiroftardis) [reply execution_engine/src/runtime/auction_internal.rs:362]: Nvm, I see us using it later on the mint transfer

File: node/src/reactor/main_reactor/tests/transactions.rs
Comment (darthsiroftardis) [inline node/src/reactor/main_reactor/tests/transactions.rs:2951]: This was me, wasn't it

File: storage/src/system/auction/detail.rs
Comment (darthsiroftardis) [inline storage/src/system/auction/detail.rs:269]: If I am reading the function body correctly, we are getting the purses for the accounts only here, perhaps we should rename the function to match that, (nit)

File: storage/src/system/auction/detail.rs
Comment (darthsiroftardis) [inline storage/src/system/auction/detail.rs]: Nit: Possible rename to `unbond_era`

File: storage/src/system/auction/detail.rs
Comment (darthsiroftardis) [inline storage/src/system/auction/detail.rs:1317]: This would only give us the accounts delegated the to the imputed validator right? Is there a reason we don't consider purses here as well?

File: storage/src/system/auction/detail.rs
Comment (darthsiroftardis) [inline storage/src/system/auction/detail.rs:1348]: Same question about only considering the accounts that have reservations

File: types/src/gens.rs
Comment (darthsiroftardis) [inline types/src/gens.rs]: Intentional?

File: utils/global-state-update-gen/src/generic/state_tracker.rs
Comment (darthsiroftardis) [inline utils/global-state-update-gen/src/generic/state_tracker.rs]: Commented code

File: utils/global-state-update-gen/src/generic/state_tracker.rs
Comment (darthsiroftardis) [inline utils/global-state-update-gen/src/generic/state_tracker.rs]: Commented code

**Review by darthsiroftardis — COMMENTED** [review]
Overall no blockers from my end over this PR, I'm waiting for the final set of commits to address CI, but assuming nothing radically changes, I can approve pending CI going green

File: types/src/gens.rs
Comment (EdHastingsCasperAssociation) [reply types/src/gens.rs]: It is now unused, in favor of delegator_bid_arb()

File: storage/src/system/auction/detail.rs
Comment (EdHastingsCasperAssociation) [reply storage/src/system/auction/detail.rs:1317]: good catch, fixed on next push

File: storage/src/system/auction/detail.rs
Comment (EdHastingsCasperAssociation) [reply storage/src/system/auction/detail.rs:1348]: Same answer.

File: storage/src/data_access_layer/auction.rs
Comment (zajko) [inline storage/src/data_access_layer/auction.rs:242]: A nit: looks like this match happens in a couple of places

**Review by darthsiroftardis — APPROVED** [review]

**Pull request merged**
